# Индивидуальное домашнее задание 3. Мусаев Умахан Рашидович.
# Вариант 23.


Зафиксируем данные:

In [34]:
import numpy as np
from scipy.stats import t
from scipy.stats import f

Y = [24.1500,25.3000,19.1894,18.2255,30.4294,40.5046,38.6738,29.5692,41.8686,47.9419,26.3229,38.0483,28.5034,45.8346,51.8930,41.2216,47.4094,56.0713,54.2849,38.7733,45.1757,59.0641,56.4441,56.5870,40.5643,57.4706,51.9329,53.0277,68.1295,51.1731,50.1222,67.0726,50.4968,60.4739,61.7605,66.2960,79.5909,69.2372,91.6667,75.5569,89.1120,91.5349,83.1951,71.1271,78.5139,86.6339,73.1428,76.8715,94.3806,82.0646,70.1522,80.2270,89.1316,88.9166,88.3489,93.6875,87.2721,80.8887,65.9675,82.6866,82.7051,91.6901,98.6271,103.3247,107.0201,108.0418,112.1984,110.6151,116.3507,116.9174,117.4782,109.0962,98.2532,84.4714,102.5327,109.7155,113.3077,112.0602,101.1875,98.9728,128.7521,127.1404,112.7526,116.5198,127.7718,139.7193,141.6801,136.3580,117.6303,127.8945,128.2121,142.3311,119.5670,131.0601,146.3962,147.2146,133.6896,129.5645,147.1280,129.5782,139.7095,122.6346,128.8803,126.7469,154.8229,159.4524,183.0954,176.4834,171.8877,156.0151,140.3092,156.4062,165.9780,165.0449,161.5379,174.6609,173.2793,155.1554,157.2397,158.3018,176.4840,176.1835,158.8950,155.6323,159.6018,153.2729,166.3936,161.4183,171.6900,177.6360,178.5838,175.5230,179.4207,184.6918,169.6380,172.2524,165.9564,167.8846,181.2260,182.5145,172.4417,167.3273,179.2495,180.9993,180.6929,191.2072,172.8856,195.8369,178.8989,195.0354,201.2496,203.2569,209.0510,205.5852,202.1229,195.2868,198.4991,210.8110,206.3392,201.1360,202.1114,217.2724,224.6388,214.4366,219.6041,218.4386,208.2906,207.5966,205.1082,202.3995,231.6358,244.4584,226.7498,234.3889,219.1299,219.9527,220.4952,207.6660,207.5717,217.3744,243.8632,240.3521,232.1431,231.4563,248.7810,243.3751,248.2755,235.1292,231.7184,251.3043,253.2731,250.2479,259.4538,255.5871,242.2632,252.4315,246.8525,238.9333,232.2496,248.3273]
X = [i for i in range(1, 201)]

# Пункт А. МНК-оценки параметров a и b:

Для нахождения МНК-оценки найдем выборочное среднее для каждой выборки:

$$ \overline{X}  = \frac{1}{n} \sum_{i = 1}^n X_i$$

$$ \overline{Y}  = \frac{1}{n} \sum_{i = 1}^n Y_i$$

In [35]:
X_srednee = sum(X)/len(X)
print("Среднее выборочное по X", X_srednee)

Y_srednee = sum(Y)/len(Y)
print("Среднее выборочное по Y", Y_srednee)



Среднее выборочное по X 100.5
Среднее выборочное по Y 139.80614350000008


Параметр b мы можем найти по формуле:

$$ b = \frac{\sum(X_i - \overline{X})(Y_i - \overline{Y})}{\sum(X_i- \overline{X})^2}$$

In [36]:
chislitel = 0
for i in range(200):
  chislitel += (X[i] - X_srednee)*(Y[i] - Y_srednee)
znaminatel = 0
for i in range(200):
  znaminatel += (X[i] - X_srednee)**2
b = chislitel/znaminatel
print("Параметр b равен:", b)

Параметр b равен: 1.128821359258982


Теперь найдем параметр а:

$$ a = \overline{Y} - b \cdot \overline{X} $$

In [37]:
a = Y_srednee - b*X_srednee
print("Параметр a равен:", a)

Параметр a равен: 26.35959689447239


Параметры найдены.

$$ \widehat{Y_i} = 26.3596 + 1.1288 \cdot X_i $$

# Пункт Б. Оценка дисперсии $\sigma^2$

Формула дисперсии:

$$ \sigma^2 = \frac{\sum_{i=1}^n(Y_i - \widehat{Y_i})^2}{n - p}$$
где p - количество коэфицентов, в нашем случае 2.

In [38]:
sigma = 0
samsa = 0
for i in range(0,200):
  samsa += (Y[i] - (a + b * X[i]))**2
sigma = samsa / 200 - 2
print("Дисперсия равна:", sigma)

Дисперсия равна: 106.23645851136934


# Пункт 3. Оценка коэффициента детерминации:

Найдем по формуле:

$$ R^2 = \frac{RSS}{TSS} = \frac{\sum(Y_i - \widehat{Y_i})^2}{\sum(Y_i - \overline{Y_i})^2} $$

In [40]:
RSS = samsa
TSS = 0
for i in range(0,200):
  TSS += (Y[i] -  Y_srednee)**2
koef_r = RSS/TSS
print("Коэффициента детерминации равен:",koef_r)

Коэффициента детерминации равен: 0.024850015686007393


Низкое значение говорит о слабой объясняющей способности модели

# Пункт 4. Точечная и интервальная (уровня надёжности 0.95) оценки для
зависимой переменной Y в точках $t_{100}$=100 и $t_{200}$=200:

Найдем точечные оценки:
$$ \widehat{Y_{100}} = a + b \cdot X_{100} $$
$$ \widehat{Y_{200}} = a + b \cdot X_{200} $$


In [8]:
Y_100 = a + b * X[99]
Y_200 = a + b * X[199]

print(Y_100, Y_200)

139.2417328203706 252.12386874626878


Теперь найдем интервальную оценку:

$$ T_{стат} = \frac{Y_{100} - \widehat{Y_{100}}}{\sigma_{y}} \sim t_{n-p}$$

$$ \sigma_{y} = \sigma_{ɛ}\cdot \sqrt{1+\frac{1}{n} + \frac{(X_{100} - \overline{X})^2}{\sum{(X_i - \overline{X})^2}}} $$

$$ T_{стат} = \frac{Y_{200} - \widehat{Y_{200}}}{\sigma_{y}} \sim t_{n-p}$$

$$ \sigma_{y} = \sigma_{ɛ}\cdot \sqrt{1+\frac{1}{n} + \frac{(X_{200} - \overline{X})^2}{\sum{(X_i - \overline{X})^2}}} $$


Интервал:

$$ (Y_{100} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}; Y_{100} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}) $$

$$ (Y_{200} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}; Y_{200} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}) $$



In [11]:

n = 200
p = 2
alpha = 0.05
sigma_ɛ = np.sqrt(sigma)
sum_X_diff_sq = znaminatel

def confidence_interval(x_new):
    y_hat = a + b * X[x_new]
    se_y = sigma_ɛ * np.sqrt(1 + 1/n + (X[x_new] - X_srednee)**2 / sum_X_diff_sq)
    t_crit = t.ppf(1 - alpha/2, df=n-p)
    lower = y_hat - t_crit * se_y
    upper = y_hat + t_crit * se_y
    return y_hat, (lower, upper)
y_100, ci_100 = confidence_interval(100-1)
y_200, ci_200 = confidence_interval(200-1)

print(f"Точка t=100:")
print(f"  Точечная оценка: {y_100:.4f}")
print(f"  95% доверительный интервал: ({ci_100[0]:.4f}, {ci_100[1]:.4f})")

print(f"\nТочка t=200:")
print(f"  Точечная оценка: {y_200:.4f}")
print(f"  95% доверительный интервал: ({ci_200[0]:.4f}, {ci_200[1]:.4f})")

Точка t=100:
  Точечная оценка: 139.2417
  95% доверительный интервал: (118.8652, 159.6183)

Точка t=200:
  Точечная оценка: 252.1239
  95% доверительный интервал: (231.5973, 272.6504)


# Пункт 5. Прогноз для значения Y в точке t=201:

Точечная оценка:

$$ Y_{201} = a + b \cdot X_{201} $$  

Интервал:

$$ (Y_{201} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}; Y_{201} - t_{1-\alpha/2,(n-p)} \cdot \sigma_{y}) $$

In [12]:
Y_201 = a + b * 201
se_Y_201 = sigma_ɛ * np.sqrt(1 + 1/n + (201 - X_srednee)**2 / sum_X_diff_sq)
t_crit = t.ppf(1 - alpha/2, df=n-p)
lower_201 = Y_201 - t_crit * se_Y_201
upper_201 = Y_201 + t_crit * se_Y_201

print("Прогноз для t=201:")
print(f"Точечная оценка: {Y_201:.4f}")
print(f"95% доверительный интервал: ({lower_201:.4f}, {upper_201:.4f})")

1.9720174778338955
Прогноз для t=201:
Точечная оценка: 253.2527
95% доверительный интервал: (232.7231, 273.7823)


---
# Часть 2. Пункт 1. Проверьте гипотезу $H_0: b = 0$:

Гипотеза:
$$ H_0: b = 0 $$
$$ H_0: b \neq 0 $$

Статистика:

$$t_0 = \frac{\widehat{b} - b}{\sigma_b}, \text{у нас b = 0} $$

$$ \sigma_b = \sigma_{ɛ} \cdot \sqrt{\frac{1}{\sum{(X_i - \overline{X})^2}}}$$

$$t_{crit} = t_{(1-\alpha/2; n-p)} $$

In [14]:
sigma_beta = sigma_ɛ * np.sqrt(1/sum_X_diff_sq)

t0 = b/sigma_beta
if (t_crit < t0):
  print("H0 отвергается")

89.42055639515297
H0 отвергается


# Пункт 2. Гипотеза о гомоскедастичности.
Упорядочиваем по возрастанию и разделяем данные.

Тестовая статистика:
$$ GK = \frac{RSS_2}{RSS_1} = \frac{3521.4}{2108.7} = 1.7945 $$
Расчет остаточных сумм квадратов для подвыборок:
$$ RSS_1 = \sum_{i=1}^k \left(Y_{} - \hat{a}_1 - \hat{b}_1 t_{i}\right)^2 $$
$$ RSS_2 = \sum_{i=n-k+1}^n \left(Y_{i} - \hat{a}_2 - \hat{b}_2 t_{i}\right)^2 $$
Критическая область:
$$ F_{crit} = F_{0.95,38,38} = 1.7167 $$
$$ 1.7945 > 1.7167 \Rightarrow\text{Гетероскедастичность присутсвует} $$

In [25]:
sorted_indices = np.argsort(X)
X_sorted = np.array(X)[sorted_indices]
Y_sorted = np.array(Y)[sorted_indices]
n = len(X)
k = int(n * 0.2)
X1, Y1 = X_sorted[:k], Y_sorted[:k]
X2, Y2 = X_sorted[-k:], Y_sorted[-k:]
def ols(X, Y):
    X_mean, Y_mean = np.mean(X), np.mean(Y)
    b = np.sum((X - X_mean) * (Y - Y_mean)) / np.sum((X - X_mean)**2)
    a = Y_mean - b * X_mean
    residuals = Y - (a + b * X)
    return np.sum(residuals**2)

RSS1 = ols(X1, Y1)
RSS2 = ols(X2, Y2)
GK = RSS2 / RSS1 if RSS2 > RSS1 else RSS1 / RSS2
df = k - 2
f_crit = f.ppf(0.95, df, df)

print(f"Статистика Голдфелда-Квандта: {GK:.4f}")
print(f"Критическое значение F({df},{df}): {f_crit:.4f}")

if GK > f_crit:
    print("Отвергаем H₀ о гомоскедастичности (гетероскедастичность присутствует)")
else:
    print("Нет оснований отвергать H₀ (гомоскедастичность)")

Статистика Голдфелда-Квандта: 1.7945
Критическое значение F(38,38): 1.7167
Отвергаем H₀ о гомоскедастичности (гетероскедастичность присутствует)


# Пункт 3. Проверка значимости квадратичного члена.

Рассмотрим расширенную модель с квадратичным членом:
$$ Y_i = a + b t_i + c t_i^2 + \epsilon_i $$

Гипотезы:

$H_0: c = 0$ (квадратичный член не нужен)

$H_1: c \neq 0$ (квадратичный член значим)


Алгоритм проверки:

1. Оцениваем полную модель:

Python код


2. Вычисляем t-статистику:
$$ t_{obs} = \frac{\hat{c}}{SE(\hat{c})} = \frac{-0.00013}{0.00025} = -0.511 $$

3. Критическое значение (α=0.05):
$$ t_{crit} = t_{0.975, 197} \approx 1.972 $$

4. Принимаем решение:
$$ |t_{obs}| = 0.511 < 1.972 \Rightarrow \text{Не отвергаем } H_0 $$

Альтернативный F-тест:

1. Вычисляем статистику:
$$ F = \frac{(RSS_{lin} - RSS_{quad})/1}{RSS_{quad}/197} = 0.261 $$

2. Критическое значение:
$$ F_{crit} = F_{0.95,1,197} \approx 3.84 $$

3. Решение:
$$ 0.261 < 3.84 \Rightarrow \text{Не отвергаем } H_0 $$

Вывод:
На уровне значимости 0.05 нет оснований включать квадратичный член в модель ($p=0.610$).

In [31]:
import statsmodels.api as sm
X_design = np.column_stack([np.ones(len(X)), X, np.square(X)])

model_full = sm.OLS(Y, X_design).fit()

c_hat = model_full.params[2]
se_c = model_full.bse[2]

t_stat = c_hat / se_c

alpha = 0.05
df = len(X) - 3
t_crit = t.ppf(1 - alpha/2, df)

print(f"Оценка коэффициента c: {c_hat:.6f}")
print(f"Стандартная ошибка: {se_c:.6f}")
print(f"t-статистика: {t_stat:.4f}")
print(f"Критическое значение t({df}): ±{t_crit:.4f}")

if abs(t_stat) > t_crit:
    print("Отвергаем H0: квадратичный член значим (p < 0.05)")
else:
    print("Не отвергаем H0: квадратичный член незначим (p ≥ 0.05)")


X_linear = np.column_stack([np.ones(len(X)), X])
model_linear = sm.OLS(Y, X_linear).fit()

RSS_full = model_full.ssr
RSS_linear = model_linear.ssr
F_stat = ((RSS_linear - RSS_full) / 1) / (RSS_full / (len(X) - 3))

F_crit = f.ppf(1 - alpha, 1, len(X) - 3)

print(f"\nF-статистика: {F_stat:.4f}")
print(f"Критическое значение F(1,{df}): {F_crit:.4f}")

if F_stat > F_crit:
    print("Отвергаем H0: квадратичный член значим (F-тест)")
else:
    print("Не отвергаем H0: квадратичный член незначим (F-тест)")

p_value = model_full.pvalues[2]
print(f"\nP-value для квадратичного члена: {p_value:.4f}")

Оценка коэффициента c: -0.000127
Стандартная ошибка: 0.000248
t-статистика: -0.5114
Критическое значение t(197): ±1.9721
Не отвергаем H0: квадратичный член незначим (p ≥ 0.05)

F-статистика: 0.2615
Критическое значение F(1,197): 3.8891
Не отвергаем H0: квадратичный член незначим (F-тест)

P-value для квадратичного члена: 0.6096
